File: Java/LinkedList/StartingOfLoop.java
/*
*
*   @author
*   Aakash Verma
*	
* 	Start of the Loop
*
*	Output: 
*
*	The list is: 3 4 5 6 7 
*	Middle Of List is: 5
*	The list is: 3 4 5 6 7 8 
*	Middle Of List is: 6
*
* 	Note: In case of even number of nodes, we are printing the ((n/2)+1)th node. 
*	If you want to print (n/2)th node you can customize your code as per your requirement.	
*
*
*/


/* Below is the structute of a node which is used to create a new node every time. */
class Node {
	int data;
	Node next;

	public Node(int key) {
		data = key;
		next = null;
	}
}

/* Creating a class for implementing the code for Starting of a loop in a Linked List. */
class StartingOfLoop {

	static Node head;

	StartingOfLoop() {
		head = null;
	}

	/* Utility code for inserting into the linked list. */
	void push(int key) {
		Node h = head;
		if(h == null) {
			Node temp = new Node(key);
			temp.next = null;
			head = temp;
		}
		else {
			while(h.next != null) {
				h = h.next;
			}
			Node temp = new Node(key);
			temp.next = null;
			h.next = temp;

		}
	}

	/* starting of a loop in the list. */
	void startingOfLoop() {
		Node fast, slow;
		fast = slow = head;
		int temp = 0;
		// boolean isLoopFound = false;
		if(head == null) {
			System.out.println("The list doesn't exists.");
			return;
		}

		while(fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			if(slow == fast) {
				temp = 1;
				// isLoopFound = true;
				break;
			}
		}

		if(temp == 1) {
			slow = head;
			while(slow != fast) {
				slow = slow.next;
				fast = fast.next;
			}
		}

		System.out.println(slow.data + " ");

	}

	/* Main method. */
	public static void main(String args[]) {
		StartingOfLoop list = new StartingOfLoop();
		list.push(3);
		list.push(4);
		list.push(5);
		list.push(6);
		list.push(7);
		list.head.next.next.next.next.next = list.head.next.next;
		list.startingOfLoop();
		System.out.println();

	}

}

File: Python/LinkedList/StartingOfLoop.py
#
#
#
#   @author
#   Aakash Verma
#	
# 	Start of a loop in Linked List
#
#	Output: 
#	5
#


# Below is the structute of a node which is used to create a new node every time.
class Node: 

	def __init__(self, data): 
		self.data = data
		self.next = None # None is nothing but null 

# Creating a class for implementing the code for Starting Of a loop in a LinkedList
class LinkedList: 
	
	# Whenever I'll create the object of a LinkedList class head will be pointing to null initially
	def __init__(self): 
		self.head = None


	# Inserting at the end of a Linked List (append function)
	def append(self, key): 

		h = self.head

		if h is None:
			new_node = Node(key)
			self.head = new_node
		else:
			while h.next != None:
				h = h.next
			new_node = Node(key)
			h.next = new_node

	# middle of a linked list
	def startingOfLoop(self):
		fast = self.head
		slow = self.head
		isLoopFound = False

		if self.head is None:
			print("The list doesn't exist.")
			return

		
		while fast is not None and fast.next is not None:
			slow = slow.next
			fast = fast.next.next
			if slow == fast:
				isLoopFound = True
				break

		if isLoopFound is True:
			slow = self.head
			while slow != fast:
				slow = slow.next
				fast = fast.next

		print(slow.data)
		


# Code execution starts here 
if __name__=='__main__': 

	myList = LinkedList() 
	myList.append(3)
	myList.append(4)
	myList.append(5)
	myList.append(6)
	myList.append(7)
	myList.head.next.next.next.next.next = myList.head.next.next
	myList.startingOfLoop()


